Re-architecture & suppliers, and tests

lordrex34 opened the pull request:
The PR contains a heavy architectural rework:

* No more SingletonHolder in ConfigManager - as it is one of the worst design concepts.
* ConfigLoadingContext is introduced.
* ConfigValueSupplier is introduced.
* Unhardcoded InputStream is introduced for override system, see test classes for example.
* PropertiesParser was replaced with ConfigProperties - which has many more powerful functions such as reading InputStream, Reader, Node (for XMLs), another ConfigProperties, or Properties, and much more.

It also fixes issues: #4 and #7

lordrex34 committed:
The beginning of another re-architecture change. :)

also: WIP #4

lordrex34 committed:
turning them to testRuntime - as Rumen suggested

lordrex34 committed:
cleanup of ConfigManager, as it was a bit crowded:

bringing back Bali's ConfigClassInfo/ConfigFieldInfo structure

lordrex34 committed:
further remodel

lordrex34 committed:
further remodel

lordrex34 committed:
minor cosmetics + allowing debug

lordrex34 committed:
further cleanup

lordrex34 committed:
further changes

lordrex34 committed:
minor changes

lordrex34 committed:
toying with post-load hooks

lordrex34 committed:
few more test helpers and such

lordrex34 committed:
Supplier test model

lordrex34 committed:
javadoc

lordrex34 committed:
better registry

lordrex34 committed:
the first normal test: override :)

lordrex34 committed:
nope, I didn't put tests into resources directory

lordrex34 committed:
minor fix

lordrex34 committed:
test for supplier

lordrex34 committed:
it really shouldn't be an issue that the config overwrites itself

lordrex34 committed:
so now testing won't be affected that much

lordrex34 committed:
javadoc fixes

lordrex34 committed:
override should be wiped also during reload

lordrex34 committed:
array config test + commenting array sorting out

lordrex34 committed:
list & set tests

Nik-- commented on src/main/java/com/github/lordrex34/config/model/ConfigFieldInfo.java:
```diff
@@ -0,0 +1,204 @@
+			// value suppliers
+			final Class<? extends IConfigValueSupplier<?>> valueSupplierClass = _configField.valueSupplier();
+			final String valueSupplierClassName = valueSupplierClass.getName();
+			IConfigValueSupplier<?> valueSupplier = VALUE_SUPPLIERS.get(valueSupplierClassName);
```
Come on... computeIfAbsent pls. There are other places like that as well.

lordrex34 replied:
```
			// value suppliers
			final Class<? extends IConfigValueSupplier<?>> valueSupplierClass = _configField.valueSupplier();
			final String valueSupplierClassName = valueSupplierClass.getName();
			IConfigValueSupplier<?> valueSupplier = VALUE_SUPPLIERS.computeIfAbsent(valueSupplierClassName, k ->
			{
				try
				{
					return valueSupplierClass.newInstance();
				}
				catch (InstantiationException | IllegalAccessException e)
				{
					e.printStackTrace();
					return null;
				}
			});
```


Feels much cleaner, huh? xD

Don't think so.

nbali replied:
with lombok it would :D

lordrex34 committed:
few more post load hooks :)

lordrex34 committed:
use "I" tag for the rest of the interfaces

lordrex34 committed:
minor change

nbali commented on build.gradle:
```diff
@@ -37,7 +37,9 @@ dependencies {
-	testCompile (group: 'junit', name: 'junit', version: '4.12')
+	testCompile(group: 'junit', name: 'junit', version: '4.12')
+	testRuntime(group: 'org.apache.logging.log4j', name: 'log4j-slf4j-impl', version: '2.8.1')
+	testRuntime(group: 'org.apache.logging.log4j', name: 'log4j-core', version: '2.8.1')
```
(technically the whole goal of slf4j is the decoupling of the actual logging implementation from a general interface... it should be included in the projects using it, not here)

lordrex34 replied:
Done.

nbali commented on .gitignore:
```diff
@@ -20,3 +20,7 @@ gradle.properties
+
+# Tests
+/config/
+/log/
```
(why would a test create any file on the disk? they should run in-memory otherwise previous runs might affect results)

lordrex34 replied:
Done.

nbali commented on src/main/java/com/github/lordrex34/config/ConfigManager.java:
```diff
@@ -77,6 +76,15 @@
 	
+	/**
+	 * Gets the configuration registry list that contains all the registered {@link ConfigClassInfo}s.
+	 * @return the configuration registry
```
(bit pointless to write the very same thing twice :D)

lordrex34 replied:
Done.

nbali commented on src/main/java/com/github/lordrex34/config/ConfigManager.java:
```diff
@@ -77,6 +76,15 @@
+	 * Gets the configuration registry list that contains all the registered {@link ConfigClassInfo}s.
+	 * @return the configuration registry
+	 */
+	public List<ConfigClassInfo> getConfigRegistry()
```
(so trustful of you to expose a simple arraylist, that isn't concurrent, nor immutable :D)

lordrex34 replied:
Done.

nbali commented on src/main/java/com/github/lordrex34/config/ConfigManager.java:
```diff
@@ -135,13 +143,22 @@ private void initOverrideProperties()
+	 * Gets overridden properties stored in this manager class.
+	 * @return overridden properties
+	 */
+	public PropertiesParser getOverriddenProperties()
```
(PropertiesParser vs get*Properties)

lordrex34 replied:
Excuse me?

nbali replied:
(it returns a PropertiesPARSER while the name indicated OverriddenPROPERTIES)

nbali commented on src/main/java/com/github/lordrex34/config/ConfigManager.java:
```diff
@@ -135,13 +143,22 @@ private void initOverrideProperties()
 	 * @param propertyKey the property key to be registered into {@code _propertiesRegistry}
 	 */
-	private void registerProperty(String packageName, Path configFile, String propertyKey)
+	public void registerProperty(String packageName, Path configFile, String propertyKey)
```
(btw puting every feature-related class into the very same package has a great advantage that you can properly control what to expose to the public and what not... for example this one makes no sense to make it visible for any end-user, and it's ony public due to package separation)

lordrex34 replied:
Done.

nbali commented on src/main/java/com/github/lordrex34/config/ConfigManager.java:
```diff
@@ -135,13 +143,22 @@ private void initOverrideProperties()
-	private void registerProperty(String packageName, Path configFile, String propertyKey)
+	public void registerProperty(String packageName, Path configFile, String propertyKey)
 	{
 		if (!_propertiesRegistry.containsKey(packageName))
```
(fyi Guava Table is what you need for easier handling :P)

nbali replied:
(or putIfAbsent)

lordrex34 replied:
Done.

nbali commented on src/main/java/com/github/lordrex34/config/ConfigManager.java:
```diff
@@ -318,22 +193,21 @@ public void load(ClassLoader classLoader, String packageName)
+		// standard annotation based configuration classes
+		try
+		{
+			ClassPathUtil.getAllClassesAnnotatedWith(classLoader, packageName, ConfigClass.class).forEach(clazz -> _configRegistry.add(new ConfigClassInfo(clazz)));
```
(I assume you control somewhere that a ```ConfigClassInfo``` isn't added for the very same class twice to the ```_configRegistry``` LIST.)

lordrex34 replied:
Done.

nbali commented on src/main/java/com/github/lordrex34/config/ConfigManager.java:
```diff
@@ -318,22 +193,21 @@ public void load(ClassLoader classLoader, String packageName)
+		catch (IOException e)
 		{
-			throw new NullPointerException("Override properties is missing!");
+			LOGGER.warn("Failed to load class path.", e);
```
(I would fail the shit out of this... or to be precise step1 that loads only from disk, try-catched with ioexception, that returns quietly, and then a step2 fail-fast parsing)

lordrex34 replied:
Done.

nbali commented on src/main/java/com/github/lordrex34/config/ConfigManager.java:
```diff
@@ -344,9 +218,9 @@ public void load(ClassLoader classLoader, String packageName)
 				
```
(not a single warning at the previous return? :D )

lordrex34 replied:
(Excuse me?)

nbali replied:
(that return; has been removed in d806391949c9c9dc9660154048584e13cd552876 )

nbali commented on src/main/java/com/github/lordrex34/config/ConfigManager.java:
```diff
@@ -358,37 +232,32 @@ public void load(ClassLoader classLoader, String packageName)
+	 * Yes. Curse of singleton holders.<br>
+	 * <b>Not to be used for reload process!</b>
+	 */
+	public void clear()
```
(just sayin', but if this is inteded to be a generally available framework, proper synchronization should be solved to prevent any misuse)

lordrex34 replied:
Removed.

nbali commented on src/main/java/com/github/lordrex34/config/ConfigManager.java:
```diff
@@ -448,13 +324,62 @@ public static Properties propertiesOf(PropertiesParser properties, PropertiesPar
+			{
+				postLoadHook = postLoadHookClass.newInstance();
+				postLoadHooks.put(postLoadHookClassName, postLoadHook);
+			}
```
(map.computeIfAbsent)

lordrex34 replied:
Done.

nbali commented on src/main/java/com/github/lordrex34/config/ConfigManager.java:
```diff
@@ -448,13 +324,62 @@ public static Properties propertiesOf(PropertiesParser properties, PropertiesPar
+		catch (InstantiationException | IllegalAccessException e)
+		{
+			LOGGER.warn("Failed to load post load hook!", e);
+		}
```
(warn only, m'kay... I would let the end-user decide what to do with this..)

nbali replied:
(this is also true for many other try-catch what wasn't modified here)

nbali commented on src/main/java/com/github/lordrex34/config/converter/ArrayConfigConverter.java:
```diff
@@ -46,10 +45,12 @@ public Object convertFromString(Field field, Class<?> type, String value)
-			Arrays.sort((Comparable[]) array);
-		}
+		// Commented out, so arrays can retain their order, and tests won't fail.
+		// I merely don't see any reason why on earth should a String array be sorted.
```
(I'm guessing determinism... but you are right, it should be a configurable option on the property... for example there are values that is easier to read in alphabetical order, sometime it isn't, and even worse the order actually matters)

Nik-- replied:
My guess was that it was done for binary search :D

lordrex34 replied:
In another feature PR I will make a boolean for sorting possibility. MAYBE.

nbali commented on src/main/java/com/github/lordrex34/config/generator/AbstractConfigGenerator.java:
```diff
@@ -60,7 +51,8 @@ public AbstractConfigGenerator(ClassLoader classLoader)
+					final ConfigClassInfo configClassInfo = new ConfigClassInfo(clazz);
+					configClassInfo.generate();
 				}
 				catch (IOException e)
```
(imo this should be fail-fast :D - and I have no idea if I implemented like this back in the day or not, but regardless fail-fast for any action like this seems more intuitive... )

lordrex34 replied:
(Excuse me?)

nbali replied:
(fail-fast -> throw the exception out... there is no reason to keep going on with the generation if one has already failed... I can call this class and have absolutely no config files generated without a single exception thrown out)

lordrex34 replied:
Done.

nbali commented on src/main/java/com/github/lordrex34/config/model/ConfigClassInfo.java:
```diff
@@ -0,0 +1,173 @@
+	private static final Logger LOGGER = LoggerFactory.getLogger(ConfigClassInfo.class);
+	
+	/** To avoid creating the same post load hook thousand times. */
+	private static final Map<String, IConfigPostLoadHook> POST_LOAD_HOOKS = new HashMap<>();
```
I assume it's static, so it's not created multiple times, but still you have two class that contains this map... why don't you create a service/manager for it then?

nbali replied:
given how many different thing uses it it could be a simple registry...  class < T > clazz, string key, supplier<T> tSupplier

nbali replied:
OR https://github.com/google/guava/wiki/CachesExplained

lordrex34 replied:
Done.

nbali commented on src/main/java/com/github/lordrex34/config/model/ConfigClassInfo.java:
```diff
@@ -0,0 +1,173 @@
+		if (_configClass == null)
+		{
+			throw new NullPointerException("Class " + _clazz + " doesn't have @ConfigClass annotation!");
+		}
```
( ```_configClass = Objects.requireNonNull(_clazz.getDeclaredAnnotation(ConfigClass.class), () -> "Class " + _clazz + " doesn't have @ConfigClass annotation!");``` )

lordrex34 replied:
Done.

nbali commented on src/main/java/com/github/lordrex34/config/model/ConfigFieldInfo.java:
```diff
@@ -0,0 +1,206 @@
+	private static final Logger LOGGER = LoggerFactory.getLogger(ConfigFieldInfo.class);
+	
+	/** To avoid creating the same value supplier thousand times. */
+	private static final Map<String, IConfigValueSupplier<?>> VALUE_SUPPLIERS = new HashMap<>();
```
(a separate service/manager/handler would be cleaner than a static which is a very bad pattern for testing)

lordrex34 replied:
Yup, Excellent idea. Done.

nbali commented on src/main/java/com/github/lordrex34/config/model/ConfigFieldInfo.java:
```diff
@@ -0,0 +1,206 @@
+		// Skip inappropriate fields.
+		if (!Modifier.isStatic(_field.getModifiers()) || Modifier.isFinal(_field.getModifiers()))
+		{
+			LOGGER.debug("Skipping non static or final field: {}#{}", _clazz.getSimpleName(), _field.getName());
```
(static final fields makes sense for constants, but a non-static field for a config class none at all... what it the use-case when we shouldn't fail immediately if that happens?)

lordrex34 replied:
(Erm.. dunno. The "original use case" can have many abnormalities I fear to comprehend... if you know what I mean.)

nbali replied:
(the original use-case contained only static fields nothing else... it could have static final for constants... both acceptable... every non-static seems concept-breaking...)

lordrex34 replied:
Done.

nbali commented on src/main/java/com/github/lordrex34/config/model/ConfigFieldInfo.java:
```diff
@@ -0,0 +1,206 @@
+			ConfigManager.loadPostLoadHook(POST_LOAD_HOOKS, _configField.postLoadHook(), properties, overriddenProperties);
+			
+			// restore field's visibility to the original
+			_field.setAccessible(wasAccessible);
```
(this should be put into a try-finally)

lordrex34 replied:
Done.

nbali commented on src/main/java/com/github/lordrex34/config/util/ClassPathUtil.java:
```diff
@@ -93,7 +93,7 @@ private ClassPathUtil()
 	 * Same as {@link #getAllClasses(ClassLoader, String)}, using {@link ClassLoader#getSystemClassLoader()} as the classLoader parameter.
 	 * @param packagePrefix
-	 * @return
+	 * @return all filtered classes
```
wasn't obvious at all... getAllClasses...

lordrex34 replied:
:D

nbali commented on src/main/java/com/github/lordrex34/config/supplier/DefaultConfigSupplier.java:
```diff
@@ -0,0 +1,78 @@
+	private static final Logger LOGGER = LoggerFactory.getLogger(DefaultConfigSupplier.class);
+	
+	/** To avoid creating the same converter thousand times. */
+	private static final Map<String, IConfigConverter> CONVERTERS = new HashMap<>();
```
(yet again static)

nbali commented on src/main/java/com/github/lordrex34/config/supplier/DefaultConfigSupplier.java:
```diff
@@ -0,0 +1,78 @@
+		catch (FieldParserException e)
+		{
+			value = converter.convertFromString(field, field.getType(), propertyValue);
+			LOGGER.warn("Property '{}' has incorrect syntax! Using default value instead: {}", propertyKey, propertyValue);
```
(I would fail here... or at least let the end-use decide)

Nik-- replied:
I agree. I would actually throw this error instead of catching it. Let the user handle it from his side.

lordrex34 replied:
Then it would break the feature requested by @UnAfraid that if user fails to specify a config properly due to whatever reason, then a default value is supposed to be used.

Remember the issues we had with "outdated configs"?

nbali replied:
given that issue it's clear that it should fail the shit out of the server... ffs... this is the worst idea like this... I create a typo and instead of the intended value it won't just fail to launch the server, but it will start with the default value? NOPE.

lordrex34 replied:
Done.

nbali commented on src/test/java/com/github/lordrex34/config/TestConfigArray.java:
```diff
@@ -0,0 +1,56 @@
+/**
+ * @author lord_rex
+ */
+public class TestConfigArray
```
(why did you separate such tests into so many classes? if you would have user @Before instead of @BeforeClass many of them could be inside the very same class as the initialization is the same)

nbali replied:
(as a general rule you should VERY rarely use @beforeclass)

lordrex34 replied:
New approach added - please re-check.

nbali commented on src/test/java/com/github/lordrex34/config/TestConfigArray.java:
```diff
@@ -0,0 +1,56 @@
+		ConfigManager.getInstance().clear();
+		
+		// load configurations
+		ConfigManager.getInstance().load(IConfigMarker.class.getPackage().getName());
```
(also singletons, static access, etc... :D not the greatest test pattern)

lordrex34 replied:
Not anymore. :)

nbali commented on src/test/java/com/github/lordrex34/config/TestConfigOverride.java:
```diff
@@ -0,0 +1,83 @@
+		try (BufferedWriter bw = Files.newBufferedWriter(overridePath))
+		{
+			bw.append(sb.toString());
+		}
```
(the framework should accept the input as string, a supplier, or at least a reader :D... or simply make an interface with writeFile/readFile and mock it here to give back what you want... it should be decoupled from the FS)

lordrex34 replied:
Done.

nbali commented on src/test/java/com/github/lordrex34/config/impl/ConfigArrayTest.java:
```diff
@@ -0,0 +1,61 @@
+ * @author lord_rex
+ */
+@ConfigClass(fileName = "array_test")
+public class ConfigArrayTest
```
and where do you empty these values between tests? :)

lordrex34 replied:
New approach added - please re-check.

nbali commented on src/test/java/com/github/lordrex34/config/impl/ConfigClassPostLoadHookTest.java:
```diff
@@ -0,0 +1,39 @@
+ * @author lord_rex
+ */
+@ConfigClass(fileName = "class_post_load_hook_test", postLoadHook = ConfigClassTestHook.class)
+public class ConfigClassPostLoadHookTest
```
and where do you empty these values between tests?

lordrex34 replied:
New approach added - please re-check.

nbali replied:
I still don't see where do you clear them... and this comment obviously goes to EVERY test that is similar to this I just have no intention to copy-pase it...

nbali commented on src/test/java/com/github/lordrex34/config/impl/ConfigFieldPostLoadHookTest.java:
```diff
@@ -0,0 +1,62 @@
+ * @author lord_rex
+ */
+@ConfigClass(fileName = "field_post_load_hook_test")
+public class ConfigFieldPostLoadHookTest
```
and where do you empty these values between tests?

lordrex34 replied:
New approach added - please re-check.

nbali commented on src/test/java/com/github/lordrex34/config/impl/ConfigListTest.java:
```diff
@@ -0,0 +1,61 @@
+ * @author lord_rex
+ */
+@ConfigClass(fileName = "list_test")
+public class ConfigListTest
```
and where do you empty these values between tests?

lordrex34 replied:
New approach added - please re-check.

nbali commented on src/test/java/com/github/lordrex34/config/impl/ConfigOverrideTest.java:
```diff
@@ -0,0 +1,38 @@
+ * @author lord_rex
+ */
+@ConfigClass(fileName = "override_test")
+public class ConfigOverrideTest
```
and where do you empty these values between tests?

lordrex34 replied:
New approach added - please re-check.

nbali commented on src/test/java/com/github/lordrex34/config/impl/ConfigSetTest.java:
```diff
@@ -0,0 +1,61 @@
+ * @author lord_rex
+ */
+@ConfigClass(fileName = "set_test")
+public class ConfigSetTest
```
and where do you empty these values between tests?

lordrex34 replied:
New approach added - please re-check.

nbali commented on src/test/java/com/github/lordrex34/config/impl/ConfigSupplierTest.java:
```diff
@@ -0,0 +1,64 @@
+ * @author lord_rex
+ */
+@ConfigClass(fileName = "supplier_test")
+public class ConfigSupplierTest
```
and where do you empty these values between tests?

lordrex34 replied:
New approach added - please re-check.

nbali commented on src/test/java/com/github/lordrex34/config/impl/ConfigTest.java:
```diff
@@ -0,0 +1,69 @@
+ * @author lord_rex
+ */
+@ConfigClass(fileName = "test")
+public class ConfigTest
```
and where do you empty these values between tests?

lordrex34 replied:
New approach added - please re-check.

nbali commented on src/test/java/com/github/lordrex34/config/impl/IConfigMarker.java:
```diff
@@ -0,0 +1,30 @@
+/**
+ * @author lord_rex
+ */
+public interface IConfigMarker
```
(ITestConfigMarker would make it easier to read in code that it isn't the same as that is used normally)

nbali commented on src/test/resources/log4j2.xml:
```diff
@@ -0,0 +1,22 @@
+			<AppenderRef ref="ConsoleFile" />
+		</Root>
+	</Loggers>
+</Configuration>
```
(as I wrote at the beginning shouldn't even be present in this project :D)

lordrex34 committed:
Merge remote-tracking branch 'origin/master' into cleanup/rearchitecture-and-suppliers

lordrex34 committed:
PR CR - tests shouldn't require logging at all

lordrex34 committed:
adding lombok as compileOnly scope

lordrex34 committed:
package marker should contain the "Test" prefix here

lordrex34 committed:
PR CR - this view should be read-only

lordrex34 committed:
this isn't used anymore

lordrex34 committed:
PR CR - moving these caches into their own class

lordrex34 committed:
PR CR - 1 liner check

lordrex34 committed:
giving properties registry it's own class

lordrex34 committed:
handling reload as static - soon you'll see why

lordrex34 committed:
saying our final good-bye to SingletonHolder in ConfigManager :)

lordrex34 committed:
PR CR - fail load if class path is not loaded properly

lordrex34 committed:
wasn't needed in the end

lordrex34 committed:
minor change:

That should be logged properly to let the user know what on earth is
happening.

lordrex34 committed:
kill this ugly design violation

lordrex34 committed:
this is supposed to be a Set

lordrex34 committed:
tests shouldn't do any disk write at all

lordrex34 committed:
that wasn't needed in the end either

lordrex34 committed:
minor change - few more explanation

lordrex34 committed:
PR CR for Nik

UnAfraid commented on src/main/java/com/github/lordrex34/config/util/ConfigPropertyRegistry.java:
```diff
@@ -0,0 +1,87 @@
+	 */
+	public static void add(String packageName, Path configFile, String propertyKey)
+	{
+		PROPERTIES.putIfAbsent(packageName, new HashMap<>());
```
Not sure why u split that into 3 lines instead of one formatted nicely

lordrex34 replied:
How would you put it into one?

UnAfraid replied:
```java
Map<String, Set<String>> entries = PROPERTIES.computeIfAbsent(packageName, key -> new HashMap<>());
entries.forEach(entry -> 
{
    final Path entryConfigFile = entry.getKey();
    final Set<String> entryProperties = entry.getValue();

    if (!entryConfigFile.equals(configFile) && entryProperties.contains(propertyKey))
    {
        LOGGER.warn("Property key '{}' is already defined in config file '{}', so now '{}' overwrites that! Please fix this!", propertyKey, entryConfigFile, configFile);
    }
});
entries.putIfAbsent(configFile, new TreeSet<>()).add(propertyKey);
```

lordrex34 replied:
Well, this is how your idea would look without compile time error:

```java
final Map<Path, Set<String>> entries = PROPERTIES.computeIfAbsent(packageName, key -> new HashMap<>());
entries.forEach((entryConfigFile, entryProperties) ->
{
	if (!entryConfigFile.equals(configFile) && entryProperties.contains(propertyKey))
	{
		LOGGER.warn("Property key '{}' is already defined in config file '{}', so now '{}' overwrites that! Please fix this!", propertyKey, entryConfigFile, configFile);
	}
});
entries.putIfAbsent(configFile, new TreeSet<>()).add(propertyKey);
```

But then the part `entries.putIfAbsent(configFile, new TreeSet<>()).add(propertyKey);` would throw a NullPointerException. :)

So?

nbali commented on src/main/java/com/github/lordrex34/config/ConfigManager.java:
```diff
@@ -57,11 +49,11 @@
-	/** A simple {@link AtomicBoolean} that indicates reloading process. */
-	private final AtomicBoolean _reloading = new AtomicBoolean(false);
+	/** Contains all the registered {@link ConfigClassInfo}s. */
+	private final Set<ConfigClassInfo> _configRegistry = new HashSet<>();
```
I don't see an equals+hashCode on the ConfigClassInfo

lordrex34 replied:
Done.

nbali commented on config/override.properties:
```diff
@@ -0,0 +1,5 @@
+# An override.properties file example.
+# Used for JUNIT tests, and also stands here as an example for the user.
+# DO NOT CHANGE ANYTHING HERE - otherwise TestConfigOverride will fail!
```
dude :D what I said about abstracting away the WHOLE file system layer? it's really not that hard :D there you have a .properties file that won't be checked by git, so any modification won't be restoreable easily, yet it could fail tests?

nbali commented on src/test/java/com/github/lordrex34/config/impl/postloadhooks/ConfigClassTestHook.java:
```diff
@@ -0,0 +1,42 @@
+/**
+ * @author lord_rex
+ */
+public class ConfigClassTestHook implements IConfigPostLoadHook
```
(btw the separation of config class, the hook, and the test itself makes it rather complicated for no real reason to see what a test does... next time just put them into a single class with inner classes...)

lordrex34 replied:
Done.

nbali commented on src/test/java/com/github/lordrex34/config/TestConfigSet.java:
```diff
@@ -0,0 +1,61 @@
+	public void before() throws IOException
+	{
+		_configManager = new ConfigManager();
+		_configManager.load(ITestConfigMarker.class.getPackage().getName());
```
(btw what I mean with the @Before instead @BeforeClass is that given that 90% of the before methods (if not all) is the same, you could have put them all into the very same file... it's so much biolerplate code like this... and before methods is executed before every method which is required for what you test..)

lordrex34 replied:
Done.

nbali commented on src/main/java/com/github/lordrex34/config/util/ConfigPropertyRegistry.java:
```diff
@@ -0,0 +1,87 @@
+	 */
+	public static void add(String packageName, Path configFile, String propertyKey)
+	{
+		PROPERTIES.putIfAbsent(packageName, new HashMap<>());
```
btw you create a new hashmap and treeset constantly for nothing... computeIfAbsent creates one ONLY if needed

lordrex34 replied:
Done.

nbali commented on src/main/java/com/github/lordrex34/config/util/ConfigPropertyRegistry.java:
```diff
@@ -0,0 +1,87 @@
+	private static final Logger LOGGER = LoggerFactory.getLogger(ConfigPropertyRegistry.class);
+	
+	/** Properties registry, that is used for misplaced configuration indication. */
+	private static final Map<String, Map<Path, Set<String>>> PROPERTIES = new TreeMap<>();
```
(btw this way this can't be tested properly... if this would have been an instance, it could have been created newly for a new test... now unless you always empty everything (which is a lot of work for nothing essentially) you will have random leftover from previous runs...)

lordrex34 replied:
Done.

nbali commented on src/main/java/com/github/lordrex34/config/util/ConfigPropertyRegistry.java:
```diff
@@ -0,0 +1,87 @@
+		PROPERTIES.putIfAbsent(packageName, new HashMap<>());
+		PROPERTIES.get(packageName).putIfAbsent(configFile, new TreeSet<>());
+		
+		PROPERTIES.get(packageName).entrySet().forEach(entry ->
```
( btw why do you only check for duplication from the same package name? why not from every package? is it because we only load from a single package at once? )

nbali replied:
anyway if this is true this whole class is pointless :D when you load a package it clears the mapping for that package, and then starts filling the mappings meanwhile checking every other mapping from the very same package... so this all could have been a single in-memory object you create when load() is called, and gets thrown away immediately after it has finished... and then all you need is a guava HashMultimap < Path, String >

lordrex34 replied:
Well, I've chosen your first option.

nbali commented on src/main/java/com/github/lordrex34/config/supplier/DefaultConfigSupplier.java:
```diff
@@ -0,0 +1,66 @@
+	public Object supply(Field field, ConfigField configField, PropertiesParser properties, PropertiesParser overridenProperties)
+	{
+		final String propertyKey = configField.name();
+		final String propertyValue = configField.value();
```
(btw looking back should have named this defaultValue() :) )

lordrex34 replied:
:D

nbali commented on src/main/java/com/github/lordrex34/config/supplier/ConfigValueSuppliers.java:
```diff
@@ -0,0 +1,66 @@
+	private static final Logger LOGGER = LoggerFactory.getLogger(ConfigValueSuppliers.class);
+	
+	/** The cache. */
+	private static final Map<String, IConfigValueSupplier<?>> VALUE_SUPPLIERS = new HashMap<>();
```
( I hope you don't test IConfigValueSuppliers :) )

lordrex34 replied:
Nope.

nbali commented on src/main/java/com/github/lordrex34/config/supplier/ConfigValueSuppliers.java:
```diff
@@ -0,0 +1,66 @@
+			catch (InstantiationException | IllegalAccessException e)
+			{
+				LOGGER.warn("Failed to load value supplier!", e);
+				return null;
```
( if this happens we will try to execute this newInstance every single time this is called... not to mention after this returns we will immediately fail with an NPE :D why don't you throw this out then? why catch is warn log it, and then fail with a non-informative exception? )

lordrex34 replied:
Done.

nbali commented on src/main/java/com/github/lordrex34/config/postloadhooks/ConfigPostLoadHooks.java:
```diff
@@ -0,0 +1,67 @@
+			catch (InstantiationException | IllegalAccessException e)
+			{
+				LOGGER.warn("Failed to load post load hook!", e);
+				return null;
```
same comments as seen at ConfigValueSuppliers.java

lordrex34 replied:
Done.

nbali commented on src/main/java/com/github/lordrex34/config/postloadhooks/ConfigPostLoadHooks.java:
```diff
@@ -0,0 +1,67 @@
+	private static final Logger LOGGER = LoggerFactory.getLogger(ConfigPostLoadHooks.class);
+	
+	/** The cache. */
+	private static final Map<String, IConfigPostLoadHook> POST_LOAD_HOOKS = new HashMap<>();
```
same comments as seen at ConfigValueSuppliers.java

lordrex34 replied:
Done.

nbali commented on src/main/java/com/github/lordrex34/config/model/ConfigFieldInfo.java:
```diff
@@ -0,0 +1,193 @@
+		}
+		catch (InstantiationException | IllegalAccessException e)
+		{
+			LOGGER.warn("Failed to set field!", e);
```
you should explain me sometime what is the point of catching every random exception and letting the load go on... this should fail the shit out of loading ... that is why we have a nice dependency graph at the loading... the config cant be loaded -> nothing loads...

lordrex34 replied:
Done.

nbali commented on src/main/java/com/github/lordrex34/config/model/ConfigFieldInfo.java:
```diff
@@ -0,0 +1,193 @@
+				}
+				
+				_field.set(null, ConfigValueSuppliers.get(_configField.valueSupplier()).supply(_field, _configField, properties, overriddenProperties));
+				ConfigPostLoadHooks.get(_configField.postLoadHook()).load(properties, overriddenProperties);
```
I don't get this.. why the whole properties/overriddenProperties is provided for the post load hook when a field has been loaded? why is this needed? why the loaded single value isn't enough?

lordrex34 replied:
The "properties" part is a must-have, as I've explained in PM. The rest is fixed.

nbali commented on src/main/java/com/github/lordrex34/config/model/ConfigClassInfo.java:
```diff
@@ -0,0 +1,160 @@
+			catch (IOException e)
+			{
+				LOGGER.warn("Failed to generate config!", e);
+			}
```
( won't it fail the shit out of the loading few lines below if the file is missing? why swallow it here if it stops it later anyway? :D )

lordrex34 replied:
Done.

nbali commented on src/main/java/com/github/lordrex34/config/converter/ConfigConverters.java:
```diff
@@ -0,0 +1,66 @@
+			catch (InstantiationException | IllegalAccessException e)
+			{
+				LOGGER.warn("Failed to load converter!", e);
+				return null;
```
same comments as seen at ConfigValueSuppliers.java

nbali commented on src/main/java/com/github/lordrex34/config/converter/ConfigConverters.java:
```diff
@@ -0,0 +1,66 @@
+	private static final Logger LOGGER = LoggerFactory.getLogger(ConfigConverters.class);
+	
+	/** The cache. */
+	private static final Map<String, IConfigConverter> CONVERTERS = new HashMap<>();
```
same comments as seen at ConfigValueSuppliers.java

nbali commented on src/main/java/com/github/lordrex34/config/supplier/ConfigValueSuppliers.java:
```diff
@@ -0,0 +1,66 @@
+ * A registry to avoid creating the same value supplier thousand times.
+ * @author lord_rex
+ */
+public final class ConfigValueSuppliers
```
( btw you literally implemented the very same 3 times... apart from names, and a method call it's all the same... should have been made generic and/or shared superclass/class )

lordrex34 replied:
Done.

nbali commented on src/main/java/com/github/lordrex34/config/ConfigManager.java:
```diff
@@ -427,9 +220,9 @@ public void reload(String packageName)
+	public static boolean isReloading()
 	{
-		return _reloading.get();
+		return RELOADING.get();
```
(this killed the possibility of concurrent test execution :D)

nbali commented on src/main/java/com/github/lordrex34/config/ConfigManager.java:
```diff
@@ -57,11 +49,11 @@
-	/** Properties registry, that is used for misplaced configuration indication. */
-	private final Map<String, Map<Path, Set<String>>> _propertiesRegistry = new TreeMap<>();
+	/** A simple {@link AtomicBoolean} that indicates reloading process. */
+	private static final AtomicBoolean RELOADING = new AtomicBoolean(false);
```
what? static? why? you can have multiple config managers, and they are all bound by this? 

(not to mention this isn't proper synchronization I talked about :D I mean what should be handled is reloading from two threads at the same time :D )

lordrex34 replied:
Done.

nbali reviewed (dismissed):
.

lordrex34 committed:
PR CR - everything to be thrown

lordrex34 committed:
let's fail inappropriate fields

lordrex34 committed:
generate equals/hashCode for model classes

lordrex34 committed:
PR CR - introducing ConfigComponent system

lordrex34 committed:
PR CR - cleanup hooks, converters, suppliers

lordrex34 committed:
minor change:

re-animating the original Properties loader that was made for this
engine

lordrex34 committed:
PM CR - api cleanup

lordrex34 committed:
javadoc fix

lordrex34 committed:
introducing config loading context

lordrex34 committed:
PR CR - minor changes:

* InputStream is now configurable inside the manager, see tests for
example
* TestConfigOverride is now using an in-memory InputStream instead of
the config/override.properties which is now deleted
* default implementation provided for the override system also
* the entire config directory is ignored now as well
* ignoring out directory as well

lordrex34 committed:
PR CR - readable test classes

lordrex34 committed:
test abstractization

lordrex34 committed:
a missed annotation :)

lordrex34 committed:
satisfying findBugs plugin as well

lordrex34 committed:
satisfying codacy as well

nbali reviewed (commented):
It hasn't been addressed properly what I asked, but who am I to block the future.

lordrex34 committed:
properly addressing Bali's request regarding clearing tests
